### PR TITLE
Magic accessors begone

### PR DESCRIPTION
**Pull request by Neloop**

### CODE REVIEW

File: app/model/entity/ExerciseTag.php [hunk 1]
```diff
@@ -56,4 +52,16 @@ public function __construct(string $name, User $author, Exercise $exercise)
         $this->exercise = $exercise;
         $this->createdNow();
     }
+
+    ////////////////////////////////////////////////////////////////////////////
+
+    public function getId(): ?string
```
Comment (krulis-martin) [inline app/model/entity/ExerciseTag.php:58]: ID is int in this entity!

File: app/model/entity/ExternalLogin.php [hunk 2]
```diff
@@ -46,4 +40,26 @@ public function __construct(User $user, string $authService, string $externalId)
         $this->authService = $authService;
         $this->externalId = $externalId;
     }
+
+    ////////////////////////////////////////////////////////////////////////////
+
+    public function getId()
```
Comment (krulis-martin) [inline app/model/entity/ExternalLogin.php:46]: Type hint for return value is missing.

File: app/model/repository/AssignmentSolutionSubmissions.php [hunk 3]
```diff
@@ -8,7 +8,7 @@
 use Doctrine\ORM\EntityManagerInterface;
 
 /**
- * @method AssignmentSolutionSubmission findOrThrow($id)
+ * @extends BaseRepository<AssignmentSolutionSubmission>
```
Comment (krulis-martin) [inline app/model/repository/AssignmentSolutionSubmissions.php:11]: What is this new develrye?

File: app/model/repository/BaseRepository.php [hunk 4]
```diff
@@ -4,11 +4,16 @@
...
 use DateTime;
+use Doctrine\Common\Collections\Collection;
 use Doctrine\Common\Collections\Criteria;
 use Doctrine\ORM\EntityManagerInterface;
 use Doctrine\ORM\EntityRepository;
+use Doctrine\ORM\QueryBuilder;
 use Nette;
 
+/**
+ * @template T
```
Comment (krulis-martin) [inline app/model/repository/BaseRepository.php:15]: Smells like C++, yet we are in PHP.

File: app/model/repository/Groups.php [hunk 5]
```diff
@@ -26,13 +26,16 @@ public function __construct(EntityManagerInterface $em)
     /**
      * Find all groups belonging to specified instance.
      * @param Instance $instance
-     * @return array
+     * @return Group[]
      */
     public function findAllByInstance(Instance $instance)
```
Comment (krulis-martin) [inline app/model/repository/Groups.php:31]: Wouldn't it be better to also add `array` type hint for the return value? (same goes for some functions below in this file)

File: app/model/repository/SisValidTerms.php [hunk 6]
```diff
@@ -15,6 +15,11 @@ public function __construct(EntityManagerInterface $em)
         parent::__construct($em, SisValidTerm::class);
     }
 
+    /**
+     * @param int $year
+     * @param int $term
+     * @return SisValidTerm|null
+     */
     public function isValid($year, $term)
```
Comment (krulis-martin) [inline app/model/repository/SisValidTerms.php:23]: return value type hint

File: app/model/entity/ExerciseTag.php (hunk 1 above)
Comment (Neloop) [reply app/model/entity/ExerciseTag.php:58]: good catch

File: app/model/repository/AssignmentSolutionSubmissions.php (hunk 3 above)
Comment (Neloop) [reply app/model/repository/AssignmentSolutionSubmissions.php:11]: Basically generics for PHP, phpstan supports this so we can easily type methods in base repository by applying extends on all child repositories

See https://phpstan.org/blog/generics-in-php-using-phpdocs

File: app/model/repository/BaseRepository.php (hunk 4 above)
Comment (Neloop) [reply app/model/repository/BaseRepository.php:15]: https://phpstan.org/blog/generics-in-php-using-phpdocs

File: app/model/repository/Groups.php (hunk 5 above)
Comment (Neloop) [reply app/model/repository/Groups.php:31]: Ok, will add

**Review by krulis-martin — APPROVED** [review]

**Pull request merged**